src/java/org/apache/cassandra/utils/bytecomparable/ByteSourceInverse.java

```diff
@@ -0,0 +1,400 @@
+import java.util.Arrays;
+
+import org.apache.cassandra.db.marshal.ValueAccessor;
+
+/**
+ * Contains inverse transformation utilities for {@link ByteSource}s.
+ *
+ * See ByteComparable.md for details about the encoding scheme.
+ */
+public final class ByteSourceInverse
+{
+    private static final int INITIAL_BUFFER_CAPACITY = 32;
+
+    public static long getUnsignedFixedLengthAsLong(ByteSource byteSource, int length)
+    {
+        if (byteSource == null)
+            throw new IllegalArgumentException("Unexpected null ByteSource");
+        if (length < 1 || length > 8)
+            throw new IllegalArgumentException("Between 1 and 8 bytes can be read at a time");
+
+        long result = 0;
+        for (int i = 0; i < length; ++i)
+        {
+            int data = byteSource.next();
+            if (data == ByteSource.END_OF_STREAM)
+                throw new IllegalArgumentException(
+                        String.format("Unexpected end of stream reached after %d bytes (expected >= %d)", i, length));
+            assert data >= 0 && data <= 255;
+            result = (result << 8) | data;
+        }
+        return result;
+    }
+
+    public static <V> V getSignedFixedLength(ValueAccessor<V> accessor, ByteSource byteSource, int length)
+    {
+        if (byteSource == null)
+            throw new IllegalArgumentException("Unexpected null ByteSource");
+        if (length < 1)
+            throw new IllegalArgumentException("At least 1 byte should be read");
+
+        V result = accessor.allocate(length);
+        accessor.putByte(result, 0, (byte) (byteSource.next() ^ 0x80));
+        for (int i = 1; i < length; ++i)
+        {
+            int data = byteSource.next();
+            if (data == ByteSource.END_OF_STREAM)
+                throw new IllegalArgumentException(
+                        String.format("Unexpected end of stream reached after %d bytes (expected >= %d)", i, length));
+            assert data >= 0 && data <= 255;
+            accessor.putByte(result, i, (byte) data);
+        }
+        return result;
+    }
+
+    public static <V> V getOptionalSignedFixedLength(ValueAccessor<V> accessor, ByteSource byteSource, int length)
+    {
+        return byteSource == null ? accessor.empty() : getSignedFixedLength(accessor, byteSource, length);
+    }
+
+    public static <V> V getSignedFixedLengthFloat(ValueAccessor<V> accessor, ByteSource byteSource, int length)
+    {
+        if (byteSource == null)
+            throw new IllegalArgumentException("Unexpected null ByteSource");
+        if (length < 1)
+            throw new IllegalArgumentException("At least 1 byte should be read");
+
+        V result = accessor.allocate(length);
+        int first = byteSource.next() & 0xFF;
+        int xor = 0;
+        if (first < 0x80) // the encoding puts negative numbers below positives
+        {
+            xor = 0xFF; // invert all bits
+            first ^= xor;
+        }
+        else
+            first ^= 0x80;
```

Comment:
nit: Maybe add a comment that this inverts just the sign bit.

P.S. While it's bad to have comments just repeating what the code does, in more intricate cases like bitwise logic mixed with conditionals, I think it makes sense to put the algorithm in the comments, implementing each explained step with code under the explanation. E.g. without that approach, I think the `DecimalType` decoding will be very hard to understand just by reading the code. It's much easier with `DoubleType` but still worth the trouble IMHO.

Reply:
Added javadoc corresponding to the doc in the forward conversion, and also expanded the comments here.